<a href="https://colab.research.google.com/github/remotjoblander/price-predictor/blob/main/Land_Price_Predictor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
csv_file_path = '/content/house-prices.csv'

try:
    df = pd.read_csv(csv_file_path)
    print(f"Successfully loaded data from {csv_file_path}")
    # Display the first 5 rows of the DataFrame
    display(df.head())
except FileNotFoundError:
    print(f"Error: The file '{csv_file_path}' was not found. Please check the path.")
except Exception as e:
    print(f"An error occurred: {e}")


Successfully loaded data from /content/house-prices.csv


,Home,Price,SqFt,Bedrooms,Bathrooms,Offers,Brick,Neighborhood
0,1,114300,1790,2,2,2,No,East
1,2,114200,2030,4,2,3,No,East
2,3,114800,1740,3,2,1,No,East
3,4,94700,1980,3,2,3,No,East
4,5,119800,2130,3,3,3,No,East


### Cell 1: Load Data with Pandas

This cell is responsible for loading your data from a CSV (Comma Separated Values) file into something called a DataFrame, which is like a table in a spreadsheet. It also includes error handling in case the file isn't found or something else goes wrong.

```python
import pandas as pd
csv_file_path = '/content/house-prices.csv'

try:
    df = pd.read_csv(csv_file_path)
    print(f"Successfully loaded data from {csv_file_path}")
    # Display the first 5 rows of the DataFrame
    display(df.head())
except FileNotFoundError:
    print(f"Error: The file '{csv_file_path}' was not found. Please check the path.")
except Exception as e:
    print(f"An error occurred: {e}")
```

**Explanation:**

*   `import pandas as pd`: This line imports the `pandas` library, which is a powerful tool for working with data in Python. We give it the shorter alias `pd` so we don't have to type `pandas` every time.
*   `csv_file_path = '/content/house-prices.csv'`: This creates a variable named `csv_file_path` and stores the location (path) of your CSV file in it. **You would change this path if your file is located elsewhere.**
*   `try...except`: This is a fundamental concept for error handling.
    *   The code inside the `try` block is executed first. If an error occurs there, Python immediately jumps to the `except` block that matches the error type.
    *   `df = pd.read_csv(csv_file_path)`: If everything goes well, this is the core command. `pd.read_csv()` is a function from the pandas library that reads the data from the specified CSV file and stores it in a DataFrame object, which we name `df` (short for DataFrame).
    *   `print(f"Successfully loaded data from {csv_file_path}")`: This line prints a confirmation message. The `f` before the string means it's an "f-string", allowing you to embed variable values (like `csv_file_path`) directly within curly braces `{}`.
    *   `display(df.head())`: After loading, `df.head()` gives you the first 5 rows of your DataFrame, which is very useful to quickly see if the data loaded correctly. `display()` is a function in environments like Colab that provides a nicely formatted output for DataFrames.
    *   `except FileNotFoundError`: This block catches a specific error that occurs if the CSV file path is incorrect or the file doesn't exist.
    *   `except Exception as e`: This is a more general catch-all for any other unexpected errors during the file loading process. `e` will contain the error message.

In [ ]:
# Target variable (y) - what we want to predict
y = df['Price'].values

# Feature variable (X) - what we use to predict (size in SqFt)
# We use double brackets to keep it as a 2D array for scikit-learn
X = df[['SqFt']].values

### Cell 2: Define Target and Feature Variables

In machine learning, we often try to predict one specific value (the 'target') based on other pieces of information (the 'features'). This cell separates your data into these two categories.

```python
# Target variable (y) - what we want to predict
y = df['Price'].values

# Feature variable (X) - what we use to predict (size in SqFt)
# We use double brackets to keep it as a 2D array for scikit-learn
X = df[['SqFt']].values
```

**Explanation:**

*   `# Target variable (y) - what we want to predict`: Lines starting with `#` are comments. They are ignored by Python and are there to explain the code to humans.
*   `y = df['Price'].values`: Here, we're selecting the 'Price' column from our DataFrame `df`. This column contains the values we want to predict. `.values` converts this column into a NumPy array, which is a common format for numerical data in Python and is often required by machine learning libraries. We assign this to the variable `y` (a common convention for the target variable).
*   `X = df[['SqFt']].values`: Here, we're selecting the 'SqFt' column. This is the feature we'll use to predict the price. Notice the double square brackets `[['SqFt']]`. This is important because machine learning models often expect features to be a 2-dimensional array (even if there's only one feature), and double brackets ensure this format. We assign this to the variable `X` (a common convention for feature variables).

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

### Cell 3: Split Data into Training and Testing Sets

Before training a machine learning model, it's crucial to split your data into two parts: a training set (to teach the model) and a testing set (to evaluate how well the model learned).

```python
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
```

**Explanation:**

*   `from sklearn.model_selection import train_test_split`: This line imports the `train_test_split` function from the `sklearn.model_selection` module. `sklearn` (Scikit-learn) is a popular machine learning library in Python.
*   `X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)`:
    *   This function takes your features (`X`) and target (`y`) as input.
    *   `test_size=0.2`: This argument specifies that 20% of your data should be used for the testing set, and the remaining 80% will be for training.
    *   `random_state=42`: This is an important argument for reproducibility. If you run the `train_test_split` function multiple times without `random_state`, you might get a different split each time. Setting it to a fixed number (like 42, a common choice) ensures that the split is always the same, which is helpful for debugging and comparing results.
    *   The function returns four new variables:
        *   `X_train`: Features for training.
        *   `X_test`: Features for testing.
        *   `y_train`: Target values for training.
        *   `y_test`: Target values for testing.

In [ ]:
from sklearn.linear_model import LinearRegression

# Initialize the model
model = LinearRegression()

# Train (fit) the model
model.fit(X_train, y_train)

# Print out the learned weight (w) and bias (b)
print(f"Learned Weight (w / slope): {model.coef_[0]:.2f}")
print(f"Learned Bias (b / intercept): {model.intercept_:.2f}")

Learned Weight (w / slope): 74.04
Learned Bias (b / intercept): -17889.36


### Cell 4: Train a Linear Regression Model

This cell initializes and trains a simple machine learning model called Linear Regression, which tries to find a straight line that best fits your data to predict house prices based on square footage.

```python
from sklearn.linear_model import LinearRegression

# Initialize the model
model = LinearRegression()

# Train (fit) the model
model.fit(X_train, y_train)

# Print out the learned weight (w) and bias (b)
print(f"Learned Weight (w / slope): {model.coef_[0]:.2f}")
print(f"Learned Bias (b / intercept): {model.intercept_:.2f}")
```

**Explanation:**

*   `from sklearn.linear_model import LinearRegression`: This imports the `LinearRegression` model from Scikit-learn's `linear_model` module.
*   `model = LinearRegression()`: This creates an "instance" of the Linear Regression model. Think of it like creating a blank blueprint for a car – it exists, but it hasn't been built or taught anything yet.
*   `model.fit(X_train, y_train)`: This is the "training" step. The `fit()` method is called on our model instance. We pass it the training features (`X_train`) and the corresponding training target values (`y_train`). The model learns the relationship between `SqFt` and `Price` from this data.
*   `print(f"Learned Weight (w / slope): {model.coef_[0]:.2f}")`: After training, a Linear Regression model learns a "weight" (also called a coefficient or slope) for each feature. `model.coef_` stores these weights. `[0]` accesses the first (and in this case, only) weight, and `:.2f` formats it to two decimal places. This weight tells you how much the price is expected to change for every one-unit increase in `SqFt`.
*   `print(f"Learned Bias (b / intercept): {model.intercept_:.2f}")`: The model also learns a "bias" (or intercept). `model.intercept_` stores this value. `:.2f` again formats it to two decimal places. The bias is the predicted price when all features (in this case, SqFt) are zero.

In [ ]:
from sklearn.metrics import mean_absolute_error, r2_score

# Make predictions on the test set
y_pred = model.predict(X_test)

# Evaluate the errors
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Absolute Error: ${mae:,.2f}")
print(f"R-squared Score: {r2:.2f}")

Mean Absolute Error: $15,945.55
R-squared Score: 0.29


### Cell 5: Evaluate the Model's Performance

After training, we need to know how well our model performs on new, unseen data. This cell uses the testing data to make predictions and then calculates common metrics to assess the model's accuracy.

```python
from sklearn.metrics import mean_absolute_error, r2_score

# Make predictions on the test set
y_pred = model.predict(X_test)

# Evaluate the errors
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Absolute Error: ${mae:,.2f}")
print(f"R-squared Score: {r2:.2f}")
```

**Explanation:**

*   `from sklearn.metrics import mean_absolute_error, r2_score`: This imports two common evaluation metrics from Scikit-learn's `metrics` module.
*   `y_pred = model.predict(X_test)`: This is the "prediction" step. We use the `predict()` method on our trained `model` and pass it the features from our **testing set** (`X_test`). The model then generates its predicted house prices for these square footages, and these predictions are stored in `y_pred`.
*   `mae = mean_absolute_error(y_test, y_pred)`: This calculates the Mean Absolute Error (MAE). MAE measures the average absolute difference between the actual prices (`y_test`) and the predicted prices (`y_pred`). A lower MAE means the predictions are, on average, closer to the actual values.
*   `r2 = r2_score(y_test, y_pred)`: This calculates the R-squared (coefficient of determination) score. R-squared indicates how much of the variance in the target variable (`Price`) can be explained by the model. It ranges from 0 to 1, where 1 means the model perfectly explains the variance, and 0 means it explains none. A higher R-squared is generally better.
*   `print(f"Mean Absolute Error: ${mae:,.2f}")`: Prints the calculated MAE, formatted to two decimal places and with a dollar sign and comma separators.
*   `print(f"R-squared Score: {r2:.2f}")`: Prints the calculated R-squared score, formatted to two decimal places. In your case, an R-squared of 0.29 suggests that only about 29% of the variability in house prices is explained by the square footage alone, which might indicate that other factors (like the number of bedrooms, bathrooms, or neighborhood) are also very important for predicting price.